File: src/main/java/com/artillexstudios/axminions/minions/MinionTypes.java
package com.artillexstudios.axminions.minions;

import com.artillexstudios.axminions.utils.LogUtils;
import it.unimi.dsi.fastutil.shorts.Short2ObjectOpenHashMap;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Collection;
import java.util.Collections;
import java.util.Locale;
import java.util.concurrent.ConcurrentHashMap;

public final class MinionTypes {
    private static final ConcurrentHashMap<String, MinionType> TYPES = new ConcurrentHashMap<>(8);
    private static final Short2ObjectOpenHashMap<MinionType> TYPE_IDS = new Short2ObjectOpenHashMap<>(8);
    private static final Collection<String> KEY_SET = Collections.unmodifiableCollection(TYPES.keySet());
    private static final Logger log = LoggerFactory.getLogger(MinionTypes.class);

    public static void register(MinionType minionType) {
        String lowercase = minionType.name().toLowerCase(Locale.ENGLISH);
        if (TYPES.containsKey(lowercase)) {
            log.error("A minion type with id {} is already registered!", lowercase);
            return;
        }

        TYPES.put(lowercase, minionType);
        TYPE_IDS.put((short) minionType.id(), minionType);
        LogUtils.debug("Put {} to {}", minionType.name(), (short) minionType.id());
    }

    public static MinionType parse(short id) {
        return TYPE_IDS.get(id);
    }

    public static MinionType parse(String minionType) {
        return TYPES.get(minionType.toLowerCase(Locale.ENGLISH));
    }

    public static void unregister(String minionType) {
        String lowercase = minionType.toLowerCase(Locale.ENGLISH);
        if (!TYPES.containsKey(lowercase)) {
            log.warn("Attempted to unregister non-existent skin {}!", lowercase);
            return;
        }

        TYPES.remove(lowercase);
    }

    public static Collection<String> types() {
        return KEY_SET;
    }
}

File: api/src/main/kotlin/com/artillexstudios/axminions/api/minions/miniontype/MinionTypes.kt
package com.artillexstudios.axminions.api.minions.miniontype

import com.artillexstudios.axminions.api.AxMinionsAPI
import com.artillexstudios.axminions.api.exception.MinionTypeAlreadyRegisteredException
import com.artillexstudios.axminions.api.utils.fastFor
import java.util.Collections
import org.bukkit.World

object MinionTypes {
    private val TYPES = hashMapOf<String, MinionType>()

    @JvmStatic
    fun register(type: MinionType): MinionType {
        if (TYPES.containsKey(type.getName())) {
            throw MinionTypeAlreadyRegisteredException("A minion with type ${type.getName()} has already been registered!")
        }

        TYPES[type.getName()] = type
        type.load()
        return type
    }

    @JvmStatic
    fun loadForWorld(world: World) {
        TYPES.fastFor { _, v ->
            AxMinionsAPI.INSTANCE.getDataHandler().loadMinionsForWorld(v, world)
        }
    }

    @JvmStatic
    fun unregister(key: String) {
        TYPES.remove(key)
    }

    @JvmStatic
    fun valueOf(name: String): MinionType? {
        return TYPES[name]
    }

    @JvmStatic
    fun getMinionTypes(): Map<String, MinionType> {
        return Collections.unmodifiableMap(TYPES)
    }
}